salt/modules/junos.py

```diff
@@ -72,6 +71,81 @@
 __proxyenabled__ = ["junos"]
 
 
+class HandleFileCopy:
+    """
+    To figure out proper path either from proxy local file system
+    or proxy cache or on master. If required, then only copy from
+    master to proxy
+
+    """
+
+    def __init__(self, path, **kwargs):
+        self._file_path = path
+        self._cached_folder = None
+        self._cached_file = None
+        self._kwargs = kwargs
+
+    def __enter__(self):
+        if self._file_path.startswith("salt://"):
+            # check if file exists in cache
+            local_cache_path = __salt__["cp.is_cached"](self._file_path)
+            if local_cache_path:
+                master_hash = __salt__["cp.hash_file"](self._file_path)
+                proxy_hash = __salt__["file.get_hash"](local_cache_path)
+                # check if hash is same, else copy newly
+                if master_hash.get("hsum") == proxy_hash:
+                    if self._kwargs:
+                        self._cached_file = salt.utils.files.mkstemp()
+                        with salt.utils.files.fopen(self._cached_file, "w") as fp:
+                            template_string = __salt__["slsutil.renderer"](
+                                path=local_cache_path,
+                                default_renderer="jinja",
+                                **self._kwargs
+                            )
+                            fp.write(template_string)
+                        return self._cached_file
+                    else:
+                        return local_cache_path
+                # continue for else part
+            self._cached_folder = tempfile.mkdtemp()
+            log.debug(
+                "Caching file {0} at {1}".format(self._file_path, self._cached_folder)
+            )
+            if self._kwargs:
+                self._cached_file = __salt__["cp.get_template"](
+                    self._file_path, self._cached_folder, **self._kwargs
+                )
+            else:
+                self._cached_file = __salt__["cp.get_file"](
+                    self._file_path, self._cached_folder
+                )
+            if self._cached_file != "":
```

Comment:
Wonder what other values self._cached_file might have other than "" and None, or a better way to know if it is usable.

Reply:
get_file and get_template return full path on proxy server (if copied from master is a susccess) else ''